### ISSUE TITLE
Recent change messes up my VS project tree, projects don't build

### ISSUE THREAD

**Issue by TurkeyMan** [premake/premake-core#533]
This swarm of files used to live under the `int` folder (you can see the folder), but now they're in the root of the project. It also doesn't build in VS anymore with many link errors.

Haven't dug into it yet, but any recent patches (more recent than a9) that affect the file-tree should be suspicious.

![image](https://cloud.githubusercontent.com/assets/1146706/16677579/c8bb910c-451a-11e6-8243-08447a6f6f46.png)

**Comment by TurkeyMan** [comment]
This patch: https://github.com/premake/premake-core/commit/a2170a4f9e64b053be62cc5dd1671480e7c92894

Breaks all our projects. Haven't looked into it, I just reverted it in our build.

**Comment by starkos** [comment]
If you find merges that break your build, and you've reasonably confirmed that the problem is in Premake and not a local customization, by all means feel free to reverse the commit and re-open the discussion on the PR. I've done this multiple times, on my own commits and on others. 

Premake gets used in so many different environments it is inevitable that some breaking commits are going to land in master from time to time.

**Comment by TurkeyMan** [comment]
So, it seems that @tvandijck's patch is incompatible with @dcourtois's qt module (https://github.com/dcourtois/premake-qt, which we make heavy use of).
Judging by the stars/forks/watches on the Qt module, it looks like it's actually the most popular premake module, and I'd say that it's bad that we merge a patch into master which breaks the most popular extension ;)

Just a though; I know that everyone hates the submodules that we already have present, but I've always argued that they act as canaries (at least for the time being since we have no reasonable test infrastructure for 3rd party extensions). I wonder if the popularity of the Qt module suggests that it should also be in core? It's a totally different kind of extension than the others so it exercises surface area that isn't touched anywhere else, so it could be a valuable canary (ie, this case).

**Comment by dcourtois** [comment]
Hi, I will have a look at the issue, but this "generated" stuff added in core in the mentioned changelist looks really similar to what my Qt added is doing. So the files probably get added twice.

As a sidenote : now that more people are actively taking part of Premake's development, that would be really nice to document such changes in the Wiki, and add a reference in the commit message. Because I can only guess what the changelist does, and trying to understand what the code does is often quite tedious, and takes longer than reading 1 or 2 good old sentences with 1 example or 2 :)

**Comment by dcourtois** [comment]
First thought : I don't like that commit, at all :p

I takes too much for granted ! It imposes a fixed vpath for generated files, and that should never happen for core functionalities (at least in my opinion) The choosen vpath is "Generated", but for instance in my projects I used a another name to make the generated folder appear as the first item in my projects hierarchy. Also if there is a project with a "Generated" folder already existing, you end up mixing your user's generated folder and the one containing the generated files automatically added by the commit ... this is definitely wrong.

Next thing : their is a commented out "addRuleFile(cfg, node)" call. I really don't want to see additional files pop in my projects ... (as a general rule, I really dislike commented out code commited in a repository ...)

And finally, this way of doing thing just don't work well with addons : for instance in Qt, to be able to know if a file needs to be preprocessed, I need to open it and scan the content for the special Q_OBJECT or Q_GADGET macros. This obviously cannot be done in the Premake5.lua file of my project : there is no such filter that would allow that (and that would be crazy)
So what the addon needs to do is that it overrides the fileconfig.addconfig method, and in this, it is able to parse the file, and add custom build rule as needed.
But at this point, it's too late, because the commit relies on the fact that the custom build rules are always set prior to the generation/baking process.

I would really appreciate this commit to be rerverted. In my opinion it has no place in the core. (and also I currently have no idea how to fix my addon to workaround this ...)

**Comment by tvandijck** [comment]
For what it is worth, It was up for review for a few weeks....
No hard feelings from my end, but these comments come of really strong towards something that could have been caught pre-commit.. I'm pretty thick skinned, so no hard feelings, but we're just trying to help where we can... Anyway, if it doesn't work and people hate it that much, revert it.. ;)

That said... the 'generated' vpath is just a dummy, it's not even used in the vcxproj, since files are actually setup as dependencies on each other, and show up as if the file that generates the code is a subfolder.. In the PR I actually displayed that... it's really nice for things like google protobufs, and we use it for several other code generated things as well... Qt for example...

It is true that this change 'broke' some of our projects too, and so we just fixed our projects.

Then lastly, the reason for this change, was to have rules act similarly to buildcommand steps that generate files that can be compiled further. Other then this change there is currently no way to have a buildcommand generate a cpp file that gets compiled into an object file that gets included in the project. For example if you have 

```lua
filter { '**.proto' }
   buildcommand { 'protoc %{file.relpath} {%file.relpath}.cc' }
```

then the .cc file does NOT get included in the project, since it doesn't exist yet, and you cannot add files to the 'files' list that you don't know might be there at some point. While it is true that you can simply say 

```lua
files { 'somefile.cpp' }
```

and premake will hapily just add that file to the vcxproj... but you cannot do 
```lua
"foreach .proto file, add a file that has the name of the output from the 'buildoutputs' from the that we added to the .proto extention"
```

For rules compileable files (.cpp, .c, .cc, .cxx, etc) are automatically compiled, and I thought it would be consistent to have the same behavior for buildcommands. Since that is the behavior you get in xcode as well...

For what it is worth... having any kind of codegeneration is really difficult with premake, and I would really like to see a much better solution to that issue...

**Comment by starkos** [comment]
I'm seeing a few issues with this one as well, including non-buildable files (object files, mostly) suddenly appearing in my source list. I think it would be a good idea to reverse it while things get sorted out.

I'd prefer not to do it right this second because I'm not on a machine where I can test before committing. If one of the admin can do it sooner, go for it, otherwise I will take a look when I get back.

(Sorry, going to be on the road and in and out a fair amount the next few weeks…doing my best…)

**Comment by tvandijck** [comment]
> I'm seeing a few issues with this one as well, including non-buildable files (object files, mostly) suddenly appearing in my source list.

That might indeed not be desirable, although I guess if we're reverting clearly none of this is desirable ;) Not sure what the right answer is, but maybe instead of reverting it, we could fix a bunch of these issues, as well as making it optional using an API.

**Comment by starkos** [comment]
> the reason for this change, was to have rules act similarly to buildcommand steps that generate files that can be compiled further

This is key, we definitely want this, and kudos to Tom and Blizzard for putting it together.

Sympathy for both sides here. It was available for review for a while, and was a nice bit of work. And sometimes it is hard to see the ways a thing can break until you break it. :p

> In my opinion it has no place in the core

I don't think this is true, preprocessing a file into a buildable result is a pretty common thing to do, and is not specific to any particular exporter. Seems exactly the kind of thing Premake ought to take care of for you.

Damien—can you keep an eye in this PR and help us shake things out to make it place nice with Qt and add-ons in general? Would be a big help…

**Comment by starkos** [comment]
> That might indeed not be desirable, although I guess if we're reverting clearly none of this is desirable ;) Not sure what the right answer is, but maybe instead of reverting it, we could fix a bunch of these issues, as well as making it optional using an API.

I'd prefer to keep this automatic if possible, since it seems like the intuitive and *right* behavior when you preprocess into a a buildable file.

I had planned to investigate and offer a fix for the non-buildable files, but it might take me a bit of time to get to it and it sounds like people have other (kind of big, from the sound of it) issues?

I'm still fully in support of having this in core, and only suggesting the revert to folks can help close the gaps, since we now appear to have everyone's attention (including mine) :)

**Comment by starkos** [comment]
Sorry, for all the comments, I missed this on first read:

> I wonder if the popularity of the Qt module suggests that it should also be in core?

I don't want to sidetrack this thread, but I would be in favor of this (to be revisited, along with all of the other core modules, when we get a module download system in place).

**Comment by tvandijck** [comment]
I'm totally on board with making further fixes, and certainly considering the havok it apparently causes... But I'd like to understand the issues, and if people have repro-cases I open to investigate and fix... I understand the screenshot manu posted above, I'm just not sure why it went that way.

**Comment by TurkeyMan** [comment]
FWIW, I'm not against functionality like this patch introduces. It's an important feature that we've needed for a long time. I appreciate the patch, it just needs more work. I think we need to revert it in the meantime, not to say that should remain that way.

I think it would be ideal for @dcourtois to be involved in massaging this patch. It needs to reach a place where the Qt module can make use of this functionality provided by core rather than implement it in the extension.

**Comment by TurkeyMan** [comment]
> As for the QT stuff... we're not using it... and have no desire to use. That said, locally we also deleted monodevelop and d from the core repo... so whether this shows up in core or not, is unimportant to us.

Sure, but the key is, `premake5 test` would run against the Qt module, and each core PR would verify it doesn't break this complex and popular addon. We shouldn't unknowingly be making changes to core that break modules. If we break modules, we need to know that we're doing it, do it deliberately, and justify that decision. The Qt module touches fairly unique surface area; that is, generating source files, managing files within the project, etc.

**Comment by dcourtois** [comment]
Hi,

First off, sorry if my comment did sound quite harsh, it wasn't meant to ! It's mostly my fault for not following the premake-core develpment closely, and not being a native english speaker. So, apologies for that :)

Now, I'll gladly follow this to try and find a solution. Here is what I think : it's a good idea to have a mean to process `.proto` (and all other special files) and have the resulting compilable files automatically added to the project. But I think it should be explicit.

So instead (that is just an idea) of using the already `buildoutput` command, maybe have a `generatedfiles` (the name is bad, it's too close in meaning to `buildoutput`, but I'm bad at finding good names) that could be used instead. And only in that case the resulting files are added to the project.
Or maybe instead, keep `buildoutput` and just add a `addoutputtoproject true` (again, maybe not the best name) command to notify premake that those files should be added ?

This way, you keep the behavior introduced in the patch, and make life easier for people who have to preprocess files, and it will let addons do what they want without conflicting ?

**Comment by tvandijck** [comment]
> Sure, but the key is, premake5 test would run against the Qt module, and each core PR would verify it doesn't break this complex and popular addon.

Isn't that an argument true for _any_ module listed here:
https://github.com/premake/premake-core/wiki/Modules

There is 19 modules listed there, of which probably only a few are still functional, either because they were made recently, or have actually been maintained..

**Comment by starkos** [comment]
I kind of feel like the process is working as intended here: modules should test regularly against the latest code, report issues, and we fix them ASAP. I'm trying to move 5.0 through alpha as fast as I can, but we're going to have to accept some breakages here and there until we get there.

Like Tom said, I don't think we can be responsible for testing (and fixing? what if we find an issue?) all of the modules on each change.

**Comment by TurkeyMan** [comment]
> Isn't that an argument true for any module listed here:
https://github.com/premake/premake-core/wiki/Modules

A lot of them are generators, or variations on similar functionality. The basis for my suggestion is that the Qt module is 1, very popular, and 2, not like any others. Test suites are about maximising surface area.

> For what it is worth, if we embedded:
https://github.com/mversluys/premake-ghp

That looks a lot like the foundation of the code we've been talking about adding in core for a long time... maybe we need to get this guy involved in the integrated solution?

> We could get rid of all other modules, and simply fetch them from github on demand.

So do we have the test script fetch and run against a bunch of modules? My argument is that we need to know when we make changed to core that break 3rd party code, otherwise you end up in a situation where all modules are basically broken all the time, and that just appears to users like premake is broken all the time. Breaking end-user code is something we need to do deliberately, not accidentally.

**Comment by TurkeyMan** [comment]
> maybe we need to get this guy involved in the integrated solution?

Oh, I see he is your colleague ;) .. sounds like it's already well on the way.
I'm excited to abandon the submodules as soon as we have a proper solution, but in the mean time, I still think we need those canaries.

**Comment by starkos** [comment]
I've opened PR #539 to reverse this commit while we sort things out. If I don't hear otherwise, and if no one beats me to it, I will merge it tomorrow morning.

I think we should move discussion of Qt testing and integration to a different issue so we can focus on this particular patch here.

I *think* that when PR #539 gets merged, Tom's original PR #518 will reopen. When that happens we can move discussion back over there.

**Comment by tvandijck** [comment]
I'm closing this issue... the 'problem' was reverted, and a new PR was opened for further discussion on this particular topic.
